### PR TITLE
WIP: TSI Series File

### PR DESCRIPTION
**Pull request by benbjohnson**
@e-dard @jwilder This is still a rough cut of the series file. It's working but the series IDs are currently `uint32` offsets so it has an upper bound of a few 10s of millions of series. I'll fix that but I wanted to get this PR up so you can take a look in case you're interested. Also, I'm still investigating the performance. It's not where I want it to be.

###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] Tests pass
- [ ] CHANGELOG.md updated
- [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [ ] [InfluxQL Spec](https://github.com/influxdata/influxdb/blob/master/influxql/README.md) updated
- [ ] Provide example syntax
- [ ] Update man page when modifying a command
- [ ] Config changes: update sample config (`etc/config.sample.toml`), server `NewDemoConfig` method, and `Diagnostics` methods reporting config settings, if necessary
- [ ] [InfluxData Documentation](https://github.com/influxdata/docs.influxdata.com): issue filed or pull request submitted \<link to issue or pull request\>

### CODE REVIEW

File: tsdb/index/tsi1/series_file.go [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+		offset += uint32(len(key))
+	}
+
+	f.hashMap = m
+
+	return nil
+}
+
+// CreateSeriesIfNotExists creates series if it doesn't exist. Returns the offset of the series.
+func (f *SeriesFile) CreateSeriesIfNotExists(name []byte, tags models.Tags, buf []byte) (offset uint32, err error) {
```
Comment (jwilder) [inline tsdb/index/tsi1/series_file.go]: Is the offset returned as the series ID stable?  Will it change each time the hash is re-indexed?  If so, then I won't be able to use it in TSM files as I need a stable ID.

Comment (jwilder) [reply tsdb/index/tsi1/series_file.go]: Also, I need the field type and field name to be part of the key as well.

File: tsdb/shard.go [hunk 2]
```diff
@@ -524,16 +524,6 @@ func (s *Shard) SeriesN() int64 {
 	return s.engine.SeriesN()
 }
 
-// SeriesSketches returns the series sketches for the shard.
-func (s *Shard) SeriesSketches() (estimator.Sketch, estimator.Sketch, error) {
```
Comment (jwilder) [inline tsdb/shard.go]: Why is this removed?

File: tsdb/index/tsi1/series_file.go [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+	// Memory map file data.
+	data, err := mmap.Map(f.path, f.MaxSize)
+	if err != nil {
+		return err
+	}
+	f.data = data
+
+	// Index all series.
+	if err := f.reindex(); err != nil {
```
Comment (jwilder) [inline tsdb/index/tsi1/series_file.go]: This looks like it's rebuilding the hash index each time the file is open?  This needs to be an on-disk hash index that we can just open and start using or startup times will be slow.

Comment (benbjohnson) [reply tsdb/index/tsi1/series_file.go]: I still need to add persistence for the hash map. I went with a simpler approach for now so I could test the ingest performance.

File: tsdb/index/tsi1/series_file.go (hunk 1 above)
Comment (benbjohnson) [reply tsdb/index/tsi1/series_file.go]: The offset is from the series file which is append-only so it will be stable. I'll add field name/type as well.

File: tsdb/shard.go (hunk 2 above)
Comment (benbjohnson) [reply tsdb/shard.go]: Do we still need series sketches if we aren't merging multiple files? It seems like we'd be able to have an accurate count at all times with a single series file.

File: tsdb/index/tsi1/series_file.go (hunk 3 above)
Comment (jwilder) [reply tsdb/index/tsi1/series_file.go]: Makes sense.

File: tsdb/index/tsi1/series_file.go (hunk 1 above)
Comment (jwilder) [reply tsdb/index/tsi1/series_file.go]: Cool

**Comment by jwilder** [comment]
This was incorporated in another PR.
